File: tests/runners/go/tests/11_relations_indexes_test.go
package polygen

import "testing"

func TestRelationsIndexesContainer(t *testing.T) {
	container := NewSchemaContainer()
	user := &User{Id: 1, Email: "author@example.com", DisplayName: "Author"}
	post := &Post{Id: 10, AuthorId: user.Id, Status: PostStatusPublished, Title: "Relations"}

	container.Users.AddRow(user)
	container.Posts.AddRow(post)

	if got := container.Posts.GetByAuthorIdStatus(PostByAuthorIdStatusKey{AuthorId: user.Id, Status: PostStatusPublished}); len(got) != 1 || got[0] != post {
		t.Fatalf("composite author/status index returned %#v", got)
	}
	if got := container.GetPostAuthor(post); got != user {
		t.Fatalf("post author navigation returned %#v", got)
	}
	if got := container.FindUserPosts(user); len(got) != 1 || got[0] != post {
		t.Fatalf("reverse user posts navigation returned %#v", got)
	}
	if err := container.ValidateOrError(); err != nil {
		t.Fatalf("expected valid relations: %v", err)
	}

	invalid := NewSchemaContainer()
	invalid.Posts.AddRow(&Post{Id: 11, AuthorId: 404, Status: PostStatusDraft, Title: "Missing author"})
	result := invalid.ValidateAll()
	if result.IsValid() {
		t.Fatalf("expected missing author foreign key to fail validation")
	}
	if result.ErrorCount() != 1 {
		t.Fatalf("expected one validation error, got %d: %s", result.ErrorCount(), result.String())
	}
}

func TestRelationsIndexesBinaryRefCompositeLookup(t *testing.T) {
	user := &User{Id: 1, Email: "author@example.com", DisplayName: "Author"}
	post := &Post{Id: 10, AuthorId: user.Id, Status: PostStatusPublished, Title: "Relations"}
	draft := &Post{Id: 11, AuthorId: user.Id, Status: PostStatusDraft, Title: "Draft"}

	data, err := SaveSchemaBinaryRefContext(&SchemaBinaryRefContainer{
		Users: []*User{user},
		Posts: []*Post{post, draft},
	})
	if err != nil {
		t.Fatalf("save binary ref: %v", err)
	}

	ctx, err := OpenSchemaBinaryRefContext(data)
	if err != nil {
		t.Fatalf("open binary ref: %v", err)
	}

	refs := ctx.Posts.FindByAuthorIdStatus(PostByAuthorIdStatusKey{AuthorId: user.Id, Status: PostStatusPublished})
	if len(refs) != 1 {
		t.Fatalf("binary ref composite author/status index returned %#v", refs)
	}
	id, err := refs[0].Id()
	if err != nil || id != post.Id {
		t.Fatalf("binary ref composite lookup id = %d, %v", id, err)
	}
	status, err := refs[0].Status()
	if err != nil || status != PostStatusPublished {
		t.Fatalf("binary ref composite lookup status = %v, %v", status, err)
	}
	if got := ctx.Posts.FindByAuthorIdStatus(PostByAuthorIdStatusKey{AuthorId: 404, Status: PostStatusPublished}); len(got) != 0 {
		t.Fatalf("missing binary ref composite key returned %#v", got)
	}
}

File: tests/runners/python/tests/11_relations_indexes_test.py
from .schema import (
    ExamplesRelationsPost,
    ExamplesRelationsPostStatus,
    ExamplesRelationsUser,
)
from .schema_binary_refs import BinaryRefDocument
from .schema_container import SchemaContainer


def test_relations_indexes_container() -> None:
    container = SchemaContainer()
    user = ExamplesRelationsUser(id=1, email="author@example.com", display_name="Author")
    post = ExamplesRelationsPost(
        id=10,
        author_id=user.id,
        status=ExamplesRelationsPostStatus.Published,
        title="Relations",
    )

    container.users.add_row(user)
    container.posts.add_row(post)

    assert container.posts.find_by_author_id_status((user.id, ExamplesRelationsPostStatus.Published)) == [post]
    assert container.get_post_author(post) is user
    assert container.find_user_posts(user) == [post]
    assert container.validate_all().is_valid()
    container.validate_or_raise()

    draft = ExamplesRelationsPost(
        id=12,
        author_id=user.id,
        status=ExamplesRelationsPostStatus.Draft,
        title="Draft",
    )
    container.posts.add_row(draft)
    document = BinaryRefDocument.from_container(container)
    refs = document.posts.find_by_author_id_status((user.id, ExamplesRelationsPostStatus.Published))
    assert len(refs) == 1
    assert refs[0].id == post.id
    assert refs[0].status == ExamplesRelationsPostStatus.Published
    assert refs[0].title == "Relations"
    assert document.posts.find_by_author_id_status((404, ExamplesRelationsPostStatus.Published)) == []

    invalid = SchemaContainer()
    invalid.posts.add_row(
        ExamplesRelationsPost(
            id=11,
            author_id=404,
            status=ExamplesRelationsPostStatus.Draft,
            title="Missing author",
        )
    )
    result = invalid.validate_all()
    assert not result.is_valid()
    assert result.error_count() == 1
    assert result.errors[0].field_name == "author_id"


test_relations_indexes_container()
